### DISCUSSION TITLE
How to set overwrite-able defaults with Tailwind via classNames utility func

### DISCUSSION THREAD

**Discussion by LinnJS** [tailwindlabs/tailwindcss#7310] [Help]
I have an Icon component that imports all SVG React components and gives the ability to set defaults for all Icon usage across the application.

I do not want to use inline styles to overwrite the tailwind classes since I do not want to allow departure from the design system.

I expected to be able to put `w-5 h-5` as a first or second argument to the classNames function shown in the TailwindUI docs. Then pass in a bigger/smaller value from the top for other use cases. But it seems Tailwind does not care about the order for the sake of specificity.

How might I be able to set a default that is overwritable from the top of the Icon component? Is this possible? Is this an anti-pattern? 

className utility func:
```
export const classNames = (...classes: string[]) => {
  return classes.filter(Boolean).join(' ');
};
```

Icon component
```
/*
  the purpose of this composable icon component is to make it easy to set defaults and enforce accessibility best practices in a repeatable way
  re: https://fontawesome.com/v4.7/accessibility/

  TODO: I want the ability to set "default" class names using Tailwind
        but apparently, order does not dictate what is more specific
        going to be communicating with the Tailwind community to figure out how to set overwrite-able defaults with Tailwind
        I could use inline styles for this but I do not want a departure from the design system.
*/
export const Icon = (props: IconProps): ReactElement | null => {
  const { label, className, name, ...rest } = props;
  const hasClassName = className ? className : '';

  return (
    <>
      {label && <span className="sr-only">{label}</span>}

      <i title={label}>
        <GetIcon aria-hidden="true" className={classNames(hasClassName, 'w-5 h-5')} data-icon={name} name={name} {...rest} />
      </i>
    </>
  );
};

```

Example of Icon default expected to be overwritten by props:

```
<Icon name="github" className="w-80 h-80" label="Sign in with Github" />
```

**Comment by bill-kerr** [comment]
I would say, in general, it's an anti-pattern to have too many "override-able" classes defined as component defaults. That being said, I do think there are certain cases where it would be nice and just saying it's an anti-pattern doesn't really solve the problem. I think the blessed approach is to either create props that control this or to just make separate components that progressively apply base styles so you can choose the one you want.

I think a default size on an icon is a good example of when it would be nice to be able to override the class without worrying about CSS ordering and without having to create a bunch of props to expose size options.

I came up with a potential solution to this. Check it out: https://github.com/bill-kerr/tailwindcss-plugin-defaults
